### ISSUE TITLE
softmax_layer.cu and cudnn_softmax_layer.cu act differently.

### ISSUE THREAD

**Issue by zhuhan1236** [BVLC/caffe#4084]
When I try to finetune AlexNet model over a specific dataset, I find that `SoftmaxWithLoss` gives a excellent result but `Softmax + MultinomialLogisticLoss` failed in this mission.

The loss layer in prototxt of `SoftmaxWithLoss` version is as follow: 
`````
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "fc8"
  bottom: "label"
  include: { phase: TRAIN }
}
``````
In this version, the test accuracy in validation set achieves 60%. But if I replace the loss layer with `Softmax + MultinomialLogisticLoss`, the test accuracy stays at 3% or 4% and never increase. The prototxt is as follow:
`````
layer {
  name: "softmax"
  type: "Softmax"
  bottom: "fc8"
  top: "softmax"
  include: { phase: TRAIN }
}
layer {
  name: "loss"
  type: "MultinomialLogisticLoss"
  bottom: "softmax"
  bottom: "label"
  include: { phase: TRAIN }
}
`````
All the other parts are the same in the above two networks. According to the document, These two networks are the same except that the first version is more numerical stable. But they lead to totally different training results. I don't know whether it's because of my incorrect using or other reason.

**Comment by zhuhan1236** [comment]
The above problem occurs because that the softmax layer is implemented by cudnn_softmax_layer.cu. If I comment `USE_CUDNN := 1` in `Makefile.config`, the `Softmax + MultinomialLogisticLoss` version can get the same result with `SoftmaxWithLoss` version.

I wonder whether there is something wrong with cudnn_softmax_layer.cu? I use cudnn of version `cudnn-7.0-v3.0`, and the caffe version  is `faba632`.

**Comment by seanbell** [comment]
`Softmax + MultinomialLogisticLoss` is very numerically unstable, so I would be surprised if it worked for any dataset.

If you want to perform softmax, comparing against arbitrary distributions, I recommend writing a new layer rather than combining two layers in a numerically unstable way.  The backprop rule for softmax + cross-entropy simplifies if you consider them together.
